### PR TITLE
Issue #3749 Fix memory leak for AsyncContextListeners with annotations

### PR DESCRIPTION
**Pull request by janbartel**
#3749

The equals() method on LifeCycleCallback was not correct, leading to multiple duplicate LifeCycleCallbacks being added to the LifeCycleCallbackCollection. 

While investigating, it became apparent that it would be possible for multiple requests to try to introspect the same class, whether it was an AsyncContextListener (which is created dynamically), or even a Servlet or Filter which was not marked loadOnStartup in web.xml. Therefore, did a few things to prevent this:

- made the AnnotationIntrospector class (which is responsible for introspecting the class of all servlet artifacts like servlets, filters, listeners etc) remember a list of classes that have already been introspected, so we never waste time re-introspecting the same class
- made the AnnotationIntrospector mutually exclusive so that only 1 thread can be introspecting at any one time

Also did:
- ensured that LifeCycleCallbackCollection and PlusDescriptor processor always refer to class by canonical name, never by just class name.
- reformatting of classes if the class was modified, and added some more comments.

### CODE REVIEW

File: jetty-annotations/src/main/java/org/eclipse/jetty/annotations/PostConstructAnnotationHandler.java [hunk 1]
```diff
@@ -73,7 +73,7 @@ public void doHandle(Class clazz)
                         return;
 
                     PostConstructCallback callback = new PostConstructCallback();
```
Comment (gregw) [inline jetty-annotations/src/main/java/org/eclipse/jetty/annotations/PostConstructAnnotationHandler.java]: As discussed, construction injection is probably better here.    Perhaps fix when merging this to jetty-10

File: jetty-plus/src/main/java/org/eclipse/jetty/plus/annotation/LifeCycleCallbackCollection.java [hunk 2]
```diff
@@ -48,10 +58,11 @@
...
         if ((callback==null) || (callback.getTargetClassName()==null))
+        {
+            if (LOG.isDebugEnabled()) LOG.debug("Ignoring empty LifeCycleCallback");
             return;
+        }
 
-        if (LOG.isDebugEnabled())
-            LOG.debug("Adding callback for class="+callback.getTargetClass()+ " on "+callback.getTarget());
         Map<String, List<LifeCycleCallback>> map = null;
         if (callback instanceof PreDestroyCallback)
```
Comment (gregw) [inline jetty-plus/src/main/java/org/eclipse/jetty/plus/annotation/LifeCycleCallbackCollection.java]: would be a bit neater to be:
```java
if (callback instanceof PreDestroyCallback)
  map = preDestroyCallbacksMap;
else if (callback instanceof PostConstructCallback)
  map = postConstructCallbacksMap;
else
  throw new IllegalArgumentException ("Unsupported lifecycle callback type: "+callback);
```

File: jetty-plus/src/main/java/org/eclipse/jetty/plus/annotation/LifeCycleCallback.java [hunk 3]
```diff
@@ -76,12 +78,16 @@ public Method getTarget()
 
     public void setTarget (String className, String methodName)
     {
+        Objects.requireNonNull(className);
+        Objects.requireNonNull(methodName);
```
Comment (sbordet) [inline jetty-plus/src/main/java/org/eclipse/jetty/plus/annotation/LifeCycleCallback.java]: Can't we use constructor injection, with `final` fields, etc.

File: jetty-plus/src/main/java/org/eclipse/jetty/plus/annotation/InjectionCollection.java [hunk 4]
```diff
@@ -32,50 +33,60 @@
...
-        
-        if (LOG.isDebugEnabled())
-            LOG.debug("Adding injection for class="+(injection.getTargetClass()+ " on a "+(injection.getTarget().getName())));
-   
-        
-        List<Injection> injections = (List<Injection>)_injectionMap.get(injection.getTargetClass().getCanonicalName());
-        if (injections==null)
+        }
+
+        String canonicalName = injection.getTargetClass().getCanonicalName();
```
[the remark this answers is not in the document]
Comment (janbartel) [reply jetty-plus/src/main/java/org/eclipse/jetty/plus/annotation/InjectionCollection.java]: used getName consistently instead.

File: jetty-plus/src/main/java/org/eclipse/jetty/plus/annotation/InjectionCollection.java [hunk 5]
```diff
@@ -32,50 +33,60 @@
...
+            injections = _injectionMap.putIfAbsent(canonicalName, new CopyOnWriteArrayList<>());
+            if (injections == null)
+                injections = _injectionMap.get(canonicalName);
+        }
+
+        if (!injections.contains(injection))
         {
-            injections = new ArrayList<Injection>();
-            _injectionMap.put(injection.getTargetClass().getCanonicalName(), injections);
+            injections.add(injection);
```
[the remark this answers is not in the document]
Comment (janbartel) [reply jetty-plus/src/main/java/org/eclipse/jetty/plus/annotation/InjectionCollection.java]: Used a set.

File: jetty-plus/src/main/java/org/eclipse/jetty/plus/annotation/LifeCycleCallback.java [hunk 6]
```diff
@@ -155,31 +161,41 @@ public Method findMethod (Package pack, Class<?> clazz, String methodName, boole
...
+    
+    @Override
+    public int hashCode()
+    {
+        int hash = 0;
+        if (_className != null)
+            hash = _className.hashCode();
+        if (_methodName != null)
+            hash ^= _methodName.hashCode();
+        return hash;
```
[the remark this answers is not in the document]
Comment (janbartel) [reply jetty-plus/src/main/java/org/eclipse/jetty/plus/annotation/LifeCycleCallback.java]: done.

File: jetty-plus/src/main/java/org/eclipse/jetty/plus/annotation/LifeCycleCallbackCollection.java [hunk 7]
```diff
@@ -48,29 +58,34 @@
...
         }
-       
+
         //don't add another callback for exactly the same method
         if (!callbacks.contains(callback))
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("Adding callback for class={} on method={}", callback.getTargetClassName(), callback.getMethodName());
             callbacks.add(callback);
+        }
```
[the remark this answers is not in the document]
Comment (janbartel) [reply jetty-plus/src/main/java/org/eclipse/jetty/plus/annotation/LifeCycleCallbackCollection.java]: Done.

File: jetty-plus/src/main/java/org/eclipse/jetty/plus/annotation/LifeCycleCallback.java (hunk 3 above)
Comment (janbartel) [reply jetty-plus/src/main/java/org/eclipse/jetty/plus/annotation/LifeCycleCallback.java]: Done.

**Comment by janbartel** [comment]
@sbordet pls re-review.

File: jetty-annotations/src/main/java/org/eclipse/jetty/annotations/AnnotationIntrospector.java [hunk 8]
```diff
@@ -19,17 +19,20 @@
...
+ * Introspects a class to find various types of
+ * annotations as defined by the servlet specification.
  */
 public class AnnotationIntrospector
-{    
-    protected List<IntrospectableAnnotationHandler> _handlers = new ArrayList<IntrospectableAnnotationHandler>();
-    
+{
+    protected final Set<Class<?>> _introspectedClasses = new HashSet<>();
+    protected final List<IntrospectableAnnotationHandler> _handlers = new ArrayList<IntrospectableAnnotationHandler>();
```
Comment (sbordet) [inline jetty-annotations/src/main/java/org/eclipse/jetty/annotations/AnnotationIntrospector.java]: Make these 2 fields `private`.

File: jetty-plus/src/main/java/org/eclipse/jetty/plus/annotation/Injection.java [hunk 9]
```diff
@@ -41,18 +42,75 @@
...
+    private final Class<?> _resourceClass;
 
 
-    public Injection ()
+    public Injection (Class<?> clazz, Field field, Class<?> resourceType, String jndiName, String mappingName)
     {
+        Objects.requireNonNull(clazz);
+        Objects.requireNonNull(field);
+        _targetClass = clazz;
+        _target = field;
```
Comment (sbordet) [inline jetty-plus/src/main/java/org/eclipse/jetty/plus/annotation/Injection.java]: These can be inlined:

```java
_targetClass = Objects.requireNonNull(clazz);
etc.
```

File: jetty-plus/src/main/java/org/eclipse/jetty/plus/annotation/LifeCycleCallback.java [hunk 10]
```diff
@@ -30,22 +31,46 @@
...
+    private final String _methodName;
 
 
-    public LifeCycleCallback()
+    public LifeCycleCallback(String className, String methodName)
     {
+        Objects.requireNonNull(className);
+        Objects.requireNonNull(methodName);
+        _className = className;
+        _methodName = methodName;
```
Comment (sbordet) [inline jetty-plus/src/main/java/org/eclipse/jetty/plus/annotation/LifeCycleCallback.java]: Inline field assignment.

File: jetty-plus/src/main/java/org/eclipse/jetty/plus/annotation/LifeCycleCallback.java [hunk 11]
```diff
@@ -155,31 +153,33 @@ public Method findMethod (Package pack, Class<?> clazz, String methodName, boole
...
-        }
-        else if (!callback.getTarget().equals(getTarget()))
+        if (!LifeCycleCallback.class.isInstance(o))
             return false;
 
-        return true;
+        LifeCycleCallback callback = (LifeCycleCallback)o;
+        
+        if (this == callback)
+            return true;
```
Comment (sbordet) [inline jetty-plus/src/main/java/org/eclipse/jetty/plus/annotation/LifeCycleCallback.java]: Move this reference equality before the check `LifeCycleCallback.class.isInstance(o)`, so it'll early return and save the `instanceof`.

File: jetty-plus/src/main/java/org/eclipse/jetty/plus/annotation/LifeCycleCallbackCollection.java [hunk 12]
```diff
@@ -47,33 +60,35 @@
      */
     public void add (LifeCycleCallback callback)
     {
-        if ((callback==null) || (callback.getTargetClassName()==null))
+        if (callback == null)
+        {
+            if (LOG.isDebugEnabled()) LOG.debug("Ignoring empty LifeCycleCallback");
```
Comment (sbordet) [inline jetty-plus/src/main/java/org/eclipse/jetty/plus/annotation/LifeCycleCallbackCollection.java]: Put `LOG.debug()` on a new line.

File: jetty-plus/src/main/java/org/eclipse/jetty/plus/annotation/RunAs.java [hunk 13]
```diff
@@ -30,24 +32,19 @@
...
-    {}
-
-
-    public void setTargetClassName (String className)
+    public RunAs(String className, String roleName)
     {
+        Objects.requireNonNull(className);
+        Objects.requireNonNull(roleName);
         _className = className;
+        _roleName = roleName;
```
Comment (sbordet) [inline jetty-plus/src/main/java/org/eclipse/jetty/plus/annotation/RunAs.java]: Inline field assignments.

**Pull request merged**